File: Python/Programmers/Level 1/체육복.py
# 그리디 : 결과는 찾는 기준에서 항상 최적의 해를 구해야함

# # set은 순서가 정해져 있지 않으므로 예를 들어, 10, [8, 10], [6,7,9] 경우에는 오답
# def solution(n, lost, reserve):
#     # lost, reserve 각각 차집합을 구함 즉, lost, reserve에 겹치는 부분 제거
#     lostSet = set(lost) - set(reserve)
#     reserveSet = set(reserve) - set(lost)
    
#     for r in reserveSet:
#         if r-1 in lostSet: # r-1이 차집합 lostSet에 속한다면
#             lostSet.remove(r-1)
#         elif r+1 in lostSet: # r+1이 차집합 lostSet에 속한다면
#             lostSet.remove(r+1)
    
#     return n - len(lostSet)

# def solution(n, lost, reserve):
#     lost.sort()
#     reserve.sort()
#     for r in reserve:
#         if r in lost: # r이 lost에 속한다면
#             lost.remove(r)
#         # 본인 체육복을 읽어버리면 본인이 입는다
#         elif r in lost and r-1 not in reserve: # r-1이 lost에 속하면서, r-1이 reserve에 속하지 않는다면
#             lost.remove(r)
#         elif r+1 in lost and r+1 not in reserve: # r+1이 lost에 속하면서, r+1이 reserve에 속하지 않는다면
#             lost.remove(r)
            
#     return n - len(lost)

def solution(n, lost, reserve):
    set_reserve, set_lost = set(reserve) - set(lost), set(lost) - set(reserve)
    
    answer = 0
    for l in set_lost:
        # remove, discard 메소드의 차이는 discard는 해당 원소가 없어도 에러는 발생시키지 않는다. 따라서 discard 메소드가 더 유연함
        # l - 1 이 set_reserve에 속한다면 discard 메소드를 이용하여 해동 원소 제거
        if (l - 1) in set_reserve: set_reserve.discard(l - 1)
        # l + 1 이 set_reserve에 속한다면 discard 메소드를 이용하여 해동 원소 제거
        elif (l + 1) in set_reserve: set_reserve.discard(l + 1)
        # 체육복을 빌릴 수 없으면 count
        else: answer += 1
    
    return n - answer

File: JavaScript/Programmers/Level 1/체육복.js
// 현재 학생이 체육복을 빌려줄 수 있는지 없는지 체크 => 그리디 알고리즘 적용
// 여벌 체육복을 가지고 온 학생이 체육복을 잃어버린 경우는 본인이 입어야 하므로 lost, reserve 배열에서 각각 차집합을 구함
// 학생 번호 -1, +1인 경우에 해당하면 체육복을 빌려줌

// function solution(n, lost, reserve) {
//     let diffLost = lost.filter(number=> !reserve.includes(number)); // lost와 reserve의 차집합을 구함
//     let diffReserve = reserve.filter(number=> !lost.includes(number)); // reserve와 lost의 차집합을 구함
    
//     diffReserve.forEach(number=> { // forEach 메소드를 사용하여 diffReserve의 모든 요소를 순회
//         if (diffLost.indexOf(number - 1) !== -1) { // number - 1이 diffLost에 있다면
//             diffLost.splice(diffLost.indexOf(number - 1), 1); // 체육복을 빌려줌
//         } else if (diffLost.indexOf(number + 1) !== -1) { // number + 1이 diffLost에 있다면
//             diffLost.splice(diffLost.indexOf(number + 1), 1); // 체육복을 빌려줌
//         }
//     });
    
//     return n - diffLost.length; // 전체 학생 수에서 빌리지 못한 학생 수를 빼면 정답
// }

// function solution(n, lost, reserve) {
//     const filteredLost = lost.filter((number) => !reserve.includes(number)).sort()
//     const filteredReserve = reserve.filter((number) => !lost.includes(number)).sort()
//     const reserveObj = Object.fromEntries(filteredReserve.map((number) => [number, true]))
    
//     let answer = n - filteredReserve.length
//     filteredLost.forEach((number) => {
//         const nextNumber = !!reserveObj[number - 1] ? number - 1 : !!reserveObj[number + 1] ? number + 1 : -1
//         if (nextNumber === -1) return
        
//         answer += 1
//         reserveObj[nextNumber] = false
//     })
    
//     return answer
// }

function solution(n, lost, reserve) {
    const answer = Array(n).fill("O")
    lost.forEach((number) => answer[number - 1] = "L")
    reserve.forEach((number) => answer[number - 1] = answer[number - 1] === "L" ? "O" : "R")
    
    for (let i = 0; i < n; i++) {
        if (answer[i] !== "L") continue
        
        if (answer[i - 1] === "R") {
            [answer[i], answer[i - 1]] = ["O", "O"]
        } else if (answer[i + 1] === "R") {
            [answer[i], answer[i + 1]] = ["O", "O"]
        }
    }
    
    return answer.filter((status) => status !== "L").length
}